xdsl/dialects/csl_/csl_stencil.py

```diff
@@ -0,0 +1,271 @@
+from collections.abc import Sequence
+from itertools import pairwise
+from typing import Annotated, cast
+
+from xdsl.dialects import builtin, memref, stencil
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, TensorType
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, Dialect, Operation, ParametrizedAttribute, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    Operand,
+    ParameterDef,
+    attr_def,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_prop_def,
+    prop_def,
+    result_def,
+)
+from xdsl.parser import AttrParser, Parser
+from xdsl.printer import Printer
+from xdsl.traits import HasAncestor, Pure
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+
+@irdl_attr_definition
+class ExchangeDeclarationAttr(ParametrizedAttribute):
+    """
+    A simplified version of dmp.exchange, from which it should be lowered
+    """
+
+    name = "csl_stencil.exchange"
+
+    neighbor_param: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(
+        self,
+        neighbor: Sequence[int] | builtin.DenseArrayBase,
+    ):
+        data_type = builtin.i64
+        super().__init__(
+            [
+                (
+                    neighbor
+                    if isinstance(neighbor, builtin.DenseArrayBase)
+                    else builtin.DenseArrayBase.from_list(data_type, neighbor)
+                ),
+            ]
+        )
+
+    @classmethod
+    def from_dmp_exch_decl_attr(cls, src: dmp.ExchangeDeclarationAttr):
+        return cls(src.neighbor)
+
+    @property
+    def neighbor(self) -> tuple[int, ...]:
+        data = self.neighbor_param.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string(f"<to {list(self.neighbor)}>")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        parser.parse_characters("<")
+        parser.parse_characters("to")
```

Comment:
I would simplify the custom syntax to `csl_stencil.direction<-1,0>` (or even `csl.direction<NORTH>`)

Reply:
It's technically a point in the stencil, meaning direction + distance. I was thinking of translating it later (when lowering to calling the API functions), in order to keep `csl_stencil.access` in line with `stencil.access`, since in the current design plan they can co-exist.